File: GPTExporterIndexerAvalonia/Helpers/TagMapGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Text.RegularExpressions.Generated;
using GPTExporterIndexerAvalonia.Services;

namespace GPTExporterIndexerAvalonia.Helpers;

public class TagMapEntry
{
    public string? Document { get; set; }
    public string? Category { get; set; }
    public string? Preview { get; set; }
    public List<string> Tags { get; set; } = [];
    public DateTime? Date { get; set; }
    public string? Title { get; set; }
    public int Line { get; set; } // This is the key field for line-level markers
    public string? Context { get; set; } // Additional context from surrounding lines
    public List<string> RelatedEntries { get; set; } = []; // Cross-references to other entries
}

public static partial class TagMapGenerator
{
    private static readonly JsonSerializerOptions s_jsonOptions = new() { WriteIndented = true };
    [GeneratedRegex(@"#(\d+)\s*[-:]\s*(.+)", RegexOptions.Compiled)]
    private static partial Regex AmandaMapPatternRegex();
    [GeneratedRegex(@"(\d{4}-\d{2}-\d{2})", RegexOptions.Compiled)]
    private static partial Regex DatePatternRegex();
    [GeneratedRegex(@"title[:\s]+(.+)", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex TitlePatternRegex();

    // Contextual patterns for capturing conversational flow
    [GeneratedRegex(@"would you like me to (log|create|document|record)", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern1Regex();
    [GeneratedRegex(@"amandamap entry \d+", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern2Regex();
    [GeneratedRegex(@"amandamap threshold", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern3Regex();
    [GeneratedRegex(@"mike[^a-zA-Z]", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern4Regex();
    [GeneratedRegex(@"onyx[^a-zA-Z]", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern5Regex();
    [GeneratedRegex(@"chatgpt[^a-zA-Z]", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern6Regex();
    [GeneratedRegex(@"let me (log|create|document|record)", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern7Regex();
    [GeneratedRegex(@"i'll (log|create|document|record)", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern8Regex();
    [GeneratedRegex(@"do you want me to", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern9Regex();
    [GeneratedRegex(@"should i (log|create|document|record)", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex ContextualPattern10Regex();

    // Marker patterns for identifying significant lines
    [GeneratedRegex(@"#\d+\s*[-:]\s*.+", RegexOptions.Compiled)]
    private static partial Regex MarkerPattern1Regex();
    [GeneratedRegex(@"^[A-Z][^.!?]*[.!?]$", RegexOptions.Compiled | RegexOptions.Multiline)]
    private static partial Regex MarkerPattern2Regex();
    [GeneratedRegex(@"\*\*[^*]+\*\*", RegexOptions.Compiled)]
    private static partial Regex MarkerPattern3Regex();
    [GeneratedRegex(@"^[A-Z][a-z]+:", RegexOptions.Compiled | RegexOptions.Multiline)]
    private static partial Regex MarkerPattern4Regex();
    [GeneratedRegex(@"Chapter\s+\w+", RegexOptions.Compiled | RegexOptions.IgnoreCase)]
    private static partial Regex MarkerPattern5Regex();
    [GeneratedRegex(@"^\d+\.\s+", RegexOptions.Compiled | RegexOptions.Multiline)]
    private static partial Regex MarkerPattern6Regex();
    [GeneratedRegex(@"^- ", RegexOptions.Compiled | RegexOptions.Multiline)]
    private static partial Regex MarkerPattern7Regex();
    
    // Contextual patterns for capturing conversational flow
    private static readonly Regex[] ContextualPatterns = [
        ContextualPattern1Regex(),
        ContextualPattern2Regex(),
        ContextualPattern3Regex(),
        ContextualPattern4Regex(),
        ContextualPattern5Regex(),
        ContextualPattern6Regex(),
        ContextualPattern7Regex(),
        ContextualPattern8Regex(),
        ContextualPattern9Regex(),
        ContextualPattern10Regex(),
    ];

    // Patterns to identify significant lines that should be marked
    private static readonly Regex[] MarkerPatterns = [
        MarkerPattern1Regex(),
        MarkerPattern2Regex(),
        MarkerPattern3Regex(),
        MarkerPattern4Regex(),
        MarkerPattern5Regex(),
        MarkerPattern6Regex(),
        MarkerPattern7Regex(),
    ];
    
    // Common AmandaMap categories and their keywords
    private static readonly Dictionary<string, string[]> CategoryKeywords = new()
    {
        ["Rituals"] = ["ritual", "ceremony", "spell", "incantation", "magic", "enchantment", "conjuration"],
        ["Thresholds"] = ["threshold", "gate", "portal", "boundary", "liminal", "transition"],
        ["Entities"] = ["entity", "spirit", "being", "creature", "presence", "manifestation"],
        ["Cosmic"] = ["cosmic", "cosmos", "universe", "galaxy", "stellar", "celestial", "astral"],
        ["Transformation"] = ["transform", "change", "metamorphosis", "evolution", "transcend"],
        ["Consciousness"] = ["consciousness", "awareness", "mind", "psychic", "mental", "cognitive"],
        ["Energy"] = ["energy", "force", "power", "vibration", "frequency", "resonance"],
        ["Time"] = ["time", "temporal", "chronos", "moment", "duration", "timeline"],
        ["Space"] = ["space", "spatial", "dimension", "realm", "plane", "location"],
        ["Technology"] = ["technology", "tech", "digital", "virtual", "cyber", "electronic"],
        ["Nature"] = ["nature", "natural", "earth", "organic", "biological", "ecological"],
        ["Philosophy"] = ["philosophy", "theory", "concept", "principle", "doctrine", "belief"],
        ["Emotional"] = ["love", "feel", "emotion", "heart", "soul", "passion", "desire"],
        ["General Insights"] = ["insight", "realization", "understanding", "awareness", "recognition"],
        ["Mike"] = ["mike"],
        ["Amandamap"] = ["amandamap", "amanda map", "amanda-map"],
        ["Onyx"] = ["onyx"],
        ["ChatGPT"] = ["chatgpt", "gpt"]
    };

    public static List<TagMapEntry> GenerateTagMap(string folderPath, IProgressService? progressService = null)
    {
        var tagMapPath = Path.Combine(folderPath, "tagmap.json");
        var entries = new List<TagMapEntry>();

        DebugLogger.Log($"TagMapGenerator: Starting tagmap generation for folder '{folderPath}'");
        progressService?.StartOperation("TagMap Generation");

        var allFiles = Directory.EnumerateFiles(folderPath, "*", SearchOption.AllDirectories).ToList();
        var supportedFiles = allFiles.Where(f => 
        {
            var ext = Path.GetExtension(f).ToLowerInvariant();
            return ext == ".txt" || ext == ".json" || ext == ".md";
        }).ToList();

        progressService?.ReportProgress(0, "Scanning files", 0, supportedFiles.Count);

        for (int i = 0; i < supportedFiles.Count; i++)
        {
            var file = supportedFiles[i];
            var ext = Path.GetExtension(file).ToLowerInvariant();
            if (ext != ".txt" && ext != ".json" && ext != ".md")
                continue;

            // Skip the index.json file itself
            if (Path.GetFileName(file).Equals("index.json", StringComparison.OrdinalIgnoreCase))
                continue;

            var relative = Path.GetRelativePath(folderPath, file);
            progressService?.ReportProgress((double)i / supportedFiles.Count * 100, $"Analyzing {Path.GetFileName(file)}", i + 1, supportedFiles.Count);
            
            var fileEntries = AnalyzeFileForContextualMarkers(file, relative);
            entries.AddRange(fileEntries);
        }

        progressService?.ReportProgress(90, "Building cross-references");

        // Build cross-references between related entries
        BuildCrossReferences(entries);

        progressService?.ReportProgress(95, "Sorting entries");

        // Sort entries by document, then by line number
        entries = entries.OrderBy(e => e.Document).ThenBy(e => e.Line).ToList();

        progressService?.ReportProgress(98, "Saving tagmap");

        // Save the tagmap
        var json = JsonSerializer.Serialize(entries, s_jsonOptions);
        File.WriteAllText(tagMapPath, json);

        DebugLogger.Log($"TagMapGenerator: Generated tagmap with {entries.Count} contextual markers");
        progressService?.CompleteOperation();

        return entries;
    }

    private static List<TagMapEntry> AnalyzeFileForContextualMarkers(string filePath, string relativePath)
    {
        var entries = new List<TagMapEntry>();
        
        try
        {
            var lines = File.ReadAllLines(filePath);
            if (lines.Length == 0)
                return entries;

            // First pass: identify all significant lines and their context
            var lineContexts = new Dictionary<int, string>();
            
            for (int lineNumber = 0; lineNumber < lines.Length; lineNumber++)
            {
                var line = lines[lineNumber];
                if (string.IsNullOrWhiteSpace(line))
                    continue;

                // Check if this line should be marked
                if (ShouldMarkLine(line))
                {
                    // Capture context from surrounding lines
                    var context = ExtractContext(lines, lineNumber, 3); // 3 lines before and after
                    lineContexts[lineNumber] = context;
                }
            }

            // Second pass: create entries with contextual awareness
            for (int lineNumber = 0; lineNumber < lines.Length; lineNumber++)
            {
                var line = lines[lineNumber];
                if (string.IsNullOrWhiteSpace(line))
                    continue;

                if (ShouldMarkLine(line) || HasContextualSignificance(line))
                {
                    var entry = new TagMapEntry
                    {
                        Document = relativePath,
                        Line = lineNumber + 1, // Convert to 1-based line numbers
                        Preview = ExtractPreview(line),
                        Category = DetermineCategory(line),
                        Tags = ExtractTags(line),
                        Date = ExtractDateFromLine(line, filePath),
                        Title = ExtractTitleFromLine(line),
                        Context = lineContexts.TryGetValue(lineNumber, out var context) ? context : null
                    };

                    entries.Add(entry);
                }
            }
        }
        catch (Exception ex)
        {
            DebugLogger.Log($"TagMapGenerator: Error analyzing file '{filePath}': {ex.Message}");
        }

        return entries;
    }

    private static string ExtractContext(string[] lines, int lineNumber, int contextLines)
    {
        var start = Math.Max(0, lineNumber - contextLines);
        var end = Math.Min(lines.Length - 1, lineNumber + contextLines);
        
        var contextLinesList = new List<string>();
        for (int i = start; i <= end; i++)
        {
            if (i == lineNumber)
                contextLinesList.Add($"**{lines[i]}**"); // Highlight the main line
            else if (!string.IsNullOrWhiteSpace(lines[i]))
                contextLinesList.Add(lines[i]);
        }
        
        return string.Join("\n", contextLinesList);
    }

    private static bool HasContextualSignificance(string line)
    {
        var trimmedLine = line.Trim().ToLowerInvariant();
        
        // Check for contextual patterns
        foreach (var pattern in ContextualPatterns)
        {
            if (pattern.IsMatch(trimmedLine))
                return true;
        }
        
        // Check for specific contextual keywords
        var contextualKeywords = new[] { "mike", "amandamap", "onyx", "chatgpt", "log", "create", "document", "record" };
        foreach (var keyword in contextualKeywords)
        {
            if (trimmedLine.Contains(keyword, StringComparison.Ordinal))
                return true;
        }
        
        return false;
    }

    private static void BuildCrossReferences(List<TagMapEntry> entries)
    {
        // Group entries by document for easier cross-referencing
        var entriesByDocument = entries.GroupBy(e => e.Document ?? string.Empty).ToDictionary(g => g.Key, g => g.ToList());
        
        foreach (var documentGroup in entriesByDocument)
        {
            var documentEntries = documentGroup.Value;
            
            // Find AmandaMap entry references and link them
            var amandamapEntries = documentEntries.Where(e => 
                e.Category == "Amandamap" || 
                e.Tags.Any(t => t.Contains("amandamap", StringComparison.OrdinalIgnoreCase)) ||
                (e.Preview?.ToLowerInvariant().Contains("amandamap") ?? false)
            ).ToList();
            
            // Find contextual mentions and link them to AmandaMap entries
            var contextualMentions = documentEntries.Where(e =>
                e.Category == "Mike" || e.Category == "Onyx" || e.Category == "ChatGPT" ||
                e.Tags.Any(t => t.Contains("mike", StringComparison.OrdinalIgnoreCase) || t.Contains("onyx", StringComparison.OrdinalIgnoreCase) || t.Contains("chatgpt", StringComparison.OrdinalIgnoreCase))
            ).ToList();
            
            // Link contextual mentions to nearby AmandaMap entries
            foreach (var mention in contextualMentions)
            {
                var nearbyEntries = amandamapEntries
                    .Where(e => Math.Abs(e.Line - mention.Line) <= 10) // Within 10 lines
                    .Select(e => $"{e.Document}:{e.Line}")
                    .ToList();
                
                mention.RelatedEntries.AddRange(nearbyEntries);
            }
            
            // Find "log/create" requests and link them to subsequent AmandaMap entries
            var logRequests = documentEntries.Where(e =>
                e.Preview?.ToLowerInvariant().Contains("log") == true ||
                e.Preview?.ToLowerInvariant().Contains("create") == true ||
                e.Preview?.ToLowerInvariant().Contains("document") == true
            ).ToList();
            
            foreach (var request in logRequests)
            {
                var subsequentEntries = amandamapEntries
                    .Where(e => e.Line > request.Line && e.Line <= request.Line + 20) // Within 20 lines after
                    .Select(e => $"{e.Document}:{e.Line}")
                    .ToList();
                
                request.RelatedEntries.AddRange(subsequentEntries);
            }
        }
    }

    private static bool ShouldMarkLine(string line)
    {
        var trimmedLine = line.Trim();
        
        // Skip empty lines
        if (string.IsNullOrWhiteSpace(trimmedLine))
            return false;
            
        // Skip very short lines (less than 10 characters)
        if (trimmedLine.Length < 10)
            return false;
            
        // Check if line matches any marker patterns
        foreach (var pattern in MarkerPatterns)
        {
            if (pattern.IsMatch(trimmedLine))
                return true;
        }
        
        // Mark lines that contain significant keywords
        var lowerLine = trimmedLine.ToLowerInvariant();
        foreach (var category in CategoryKeywords)
        {
            foreach (var keyword in category.Value)
            {
                if (lowerLine.Contains(keyword))
                    return true;
            }
        }
        
        // Mark lines that are complete sentences and have substantial content
        if (trimmedLine.Length > 50 && 
            char.IsUpper(trimmedLine[0]) && 
            (trimmedLine.EndsWith('.') || trimmedLine.EndsWith('!') || trimmedLine.EndsWith('?')))
        {
            return true;
        }
        
        return false;
    }

    private static string ExtractPreview(string line)
    {
        var trimmed = line.Trim();
        // Limit preview to 200 characters
        return trimmed.Length > 200 ? string.Concat(trimmed.AsSpan(0, 200), "...") : trimmed;
    }

    private static string? ExtractTitleFromLine(string line)
    {
        // Try to find AmandaMap entry numbers and titles
        var match = AmandaMapPatternRegex().Match(line);
        if (match.Success)
        {
            return $"#{match.Groups[1].Value} - {match.Groups[2].Value.Trim()}";
        }

        // Try to find explicit title
        var titleMatch = TitlePatternRegex().Match(line);
        if (titleMatch.Success)
        {
            return titleMatch.Groups[1].Value.Trim();
        }

        // Use first part of line as title if it's substantial
        var trimmed = line.Trim();
        if (trimmed.Length > 20)
        {
            var firstSentence = trimmed.Split('.', '!', '?')[0];
            return firstSentence.Length > 10 ? firstSentence : null;
        }

        return null;
    }

    private static DateTime? ExtractDateFromLine(string line, string filePath)
    {
        // Try to extract date from the line
        var match = DatePatternRegex().Match(line);
        if (match.Success && DateTime.TryParse(match.Value, out var date))
        {
            return date;
        }

        // Use file modification date as fallback
        return File.GetLastWriteTime(filePath);
    }

    private static string? DetermineCategory(string line)
    {
        var lowerLine = line.ToLowerInvariant();
        var categoryScores = new Dictionary<string, int>();

        foreach (var category in CategoryKeywords)
        {
            var score = 0;
            foreach (var keyword in category.Value)
            {
                score += Regex.Matches(lowerLine, $@"\b{Regex.Escape(keyword)}\b").Count;
            }
            if (score > 0)
            {
                categoryScores[category.Key] = score;
            }
        }

        return categoryScores.OrderByDescending(x => x.Value).FirstOrDefault().Key ?? "General Insights";
    }

    private static List<string> ExtractTags(string line)
    {
        var tags = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var lowerLine = line.ToLowerInvariant();

        // Extract all category keywords found
        foreach (var category in CategoryKeywords)
        {
            foreach (var keyword in category.Value)
            {
                if (Regex.IsMatch(lowerLine, $@"\b{Regex.Escape(keyword)}\b"))
                {
                    tags.Add(keyword);
                }
            }
        }

        // Extract AmandaMap entry numbers
        var entryMatches = AmandaMapPatternRegex().Matches(line);
        foreach (Match match in entryMatches)
        {
            tags.Add($"entry-{match.Groups[1].Value}");
        }

        // Extract common AmandaMap themes
        var themeKeywords = new[] { "amanda", "threshold", "cosmic", "transformation", "consciousness", "energy", "time", "space" };
        foreach (var keyword in themeKeywords)
        {
            if (lowerLine.Contains(keyword))
            {
                tags.Add(keyword);
            }
        }

        // Extract contextual tags
        if (lowerLine.Contains("mike"))
            tags.Add("mike-context");
        if (lowerLine.Contains("amandamap"))
            tags.Add("amandamap-context");
        if (lowerLine.Contains("onyx"))
            tags.Add("onyx-context");
        if (lowerLine.Contains("chatgpt"))
            tags.Add("chatgpt-context");
        if (lowerLine.Contains("log") || lowerLine.Contains("create") || lowerLine.Contains("document"))
            tags.Add("logging-request");

        return tags.OrderBy(t => t).ToList();
    }

    public static List<TagMapEntry> LoadTagMap(string folderPath)
    {
        var tagMapPath = Path.Combine(folderPath, "tagmap.json");
        if (!File.Exists(tagMapPath))
        {
            DebugLogger.Log($"TagMapGenerator: No existing tagmap found at '{tagMapPath}'");
            return [];
        }

        try
        {
            var json = File.ReadAllText(tagMapPath);
            var entries = JsonSerializer.Deserialize<List<TagMapEntry>>(json);
            DebugLogger.Log($"TagMapGenerator: Loaded existing tagmap with {entries?.Count ?? 0} entries");
            return entries ?? [];
        }
        catch (Exception ex)
        {
            DebugLogger.Log($"TagMapGenerator: Error loading tagmap: {ex.Message}");
            return [];
        }
    }

    public static void UpdateTagMap(string folderPath, IProgressService? progressService = null)
    {
        DebugLogger.Log($"TagMapGenerator: Updating tagmap for folder '{folderPath}'");
        GenerateTagMap(folderPath, progressService);
    }
} 

File: modules/tagmap_generator.py
"""
TagMap Generator Module

Backports the sophisticated tagmap generation functionality from the Avalonia app's TagMapGenerator.cs.
Provides contextual markers, cross-references, and intelligent category detection.
"""

import json
import re
import os
from pathlib import Path
from typing import Dict, List, Set, Optional, Tuple, Any
from dataclasses import dataclass, asdict
from datetime import datetime
import logging

logger = logging.getLogger(__name__)

@dataclass
class TagMapEntry:
    """Represents a single entry in the tagmap with metadata and context."""
    document: Optional[str] = None
    category: Optional[str] = None
    preview: Optional[str] = None
    tags: List[str] = None
    date: Optional[str] = None
    title: Optional[str] = None
    line: int = 0
    context: Optional[str] = None
    related_entries: List[str] = None
    
    def __post_init__(self):
        if self.tags is None:
            self.tags = []
        if self.related_entries is None:
            self.related_entries = []

class TagMapGenerator:
    """Advanced tagmap generation with contextual analysis and cross-references."""
    
    # Compiled regex patterns for pattern matching
    AMANDAMAP_PATTERN = re.compile(r'#(\d+)\s*[-:]\s*(.+)', re.IGNORECASE)
    DATE_PATTERN = re.compile(r'(\d{4}-\d{2}-\d{2})')
    TITLE_PATTERN = re.compile(r'title[:\s]+(.+)', re.IGNORECASE)
    
    # Contextual patterns for capturing conversational flow
    CONTEXTUAL_PATTERNS = [
        re.compile(r'would you like me to (log|create|document|record)', re.IGNORECASE),
        re.compile(r'amandamap entry \d+', re.IGNORECASE),
        re.compile(r'amandamap threshold', re.IGNORECASE),
        re.compile(r'mike[^a-zA-Z]', re.IGNORECASE),  # Mike mentions
        re.compile(r'onyx[^a-zA-Z]', re.IGNORECASE),  # Onyx mentions
        re.compile(r'chatgpt[^a-zA-Z]', re.IGNORECASE),  # ChatGPT mentions
        re.compile(r'let me (log|create|document|record)', re.IGNORECASE),
        re.compile(r"i'll (log|create|document|record)", re.IGNORECASE),
        re.compile(r'do you want me to', re.IGNORECASE),
        re.compile(r'should i (log|create|document|record)', re.IGNORECASE),
    ]
    
    # Patterns to identify significant lines that should be marked
    MARKER_PATTERNS = [
        re.compile(r'#\d+\s*[-:]\s*.+'),  # AmandaMap entries
        re.compile(r'^[A-Z][^.!?]*[.!?]$', re.MULTILINE),  # Complete sentences starting with capital
        re.compile(r'\*\*[^*]+\*\*'),  # Bold text
        re.compile(r'^[A-Z][a-z]+:', re.MULTILINE),  # Section headers
        re.compile(r'Chapter\s+\w+', re.IGNORECASE),  # Chapter markers
        re.compile(r'^\d+\.\s+', re.MULTILINE),  # Numbered lists
        re.compile(r'^- ', re.MULTILINE),  # Bullet points
    ]
    
    # Common AmandaMap categories and their keywords
    CATEGORY_KEYWORDS = {
        "Rituals": ["ritual", "ceremony", "spell", "incantation", "magic", "enchantment", "conjuration"],
        "Thresholds": ["threshold", "gate", "portal", "boundary", "liminal", "transition"],
        "Entities": ["entity", "spirit", "being", "creature", "presence", "manifestation"],
        "Cosmic": ["cosmic", "cosmos", "universe", "galaxy", "stellar", "celestial", "astral"],
        "Transformation": ["transform", "change", "metamorphosis", "evolution", "transcend"],
        "Consciousness": ["consciousness", "awareness", "mind", "psychic", "mental", "cognitive"],
        "Energy": ["energy", "force", "power", "vibration", "frequency", "resonance"],
        "Time": ["time", "temporal", "chronos", "moment", "duration", "timeline"],
        "Space": ["space", "spatial", "dimension", "realm", "plane", "location"],
        "Technology": ["technology", "tech", "digital", "virtual", "cyber", "electronic"],
        "Nature": ["nature", "natural", "earth", "organic", "biological", "ecological"],
        "Philosophy": ["philosophy", "theory", "concept", "principle", "doctrine", "belief"],
        "Emotional": ["love", "feel", "emotion", "heart", "soul", "passion", "desire"],
        "General Insights": ["insight", "realization", "understanding", "awareness", "recognition"],
        "Mike": ["mike"],
        "Amandamap": ["amandamap", "amanda map", "amanda-map"],
        "Onyx": ["onyx"],
        "ChatGPT": ["chatgpt", "gpt"]
    }
    
    def __init__(self):
        self.entries: List[TagMapEntry] = []
        self.cross_references: Dict[str, List[str]] = {}
    
    def generate_tagmap(
        self, 
        folder_path: Path, 
        overwrite_existing: bool = False,
        progress_callback=None
    ) -> List[TagMapEntry]:
        """
        Generate a comprehensive tagmap for all files in the folder.
        
        Args:
            folder_path: Directory to analyze
            overwrite_existing: Whether to overwrite existing tagmap
            progress_callback: Optional callback for progress updates
            
        Returns:
            List of tagmap entries
        """
        logger.info(f"Generating tagmap for folder: {folder_path}")
        
        tagmap_path = folder_path / "tagmap.json"
        entries = []
        
        # Check if tagmap exists and we're not overwriting
        if tagmap_path.exists() and not overwrite_existing:
            try:
                entries = self.load_tagmap(folder_path)
                logger.info(f"Loaded existing tagmap with {len(entries)} entries")
                return entries
            except Exception as e:
                logger.warning(f"Error loading existing tagmap: {e}")
        
        # Get all supported files
        supported_extensions = {'.txt', '.json', '.md', '.html', '.xml'}
        all_files = []
        
        for file_path in folder_path.rglob('*'):
            if file_path.is_file() and file_path.suffix.lower() in supported_extensions:
                all_files.append(file_path)
        
        logger.info(f"Found {len(all_files)} files to analyze")
        
        if progress_callback:
            progress_callback("Analyzing files", 0, len(all_files))
        
        # Process each file
        for i, file_path in enumerate(all_files):
            try:
                file_entries = self._analyze_file_for_contextual_markers(file_path, folder_path)
                entries.extend(file_entries)
                
                if progress_callback:
                    progress_callback(f"Analyzed {file_path.name}", i + 1, len(all_files))
                    
            except Exception as e:
                logger.error(f"Error analyzing file {file_path}: {e}")
        
        # Build cross-references
        self._build_cross_references(entries)
        
        # Save tagmap
        try:
            with open(tagmap_path, 'w', encoding='utf-8') as f:
                json.dump([asdict(entry) for entry in entries], f, indent=2)
            logger.info(f"Tagmap saved to {tagmap_path}")
        except Exception as e:
            logger.error(f"Error saving tagmap: {e}")
        
        self.entries = entries
        return entries
    
    def _analyze_file_for_contextual_markers(
        self, 
        file_path: Path, 
        base_path: Path
    ) -> List[TagMapEntry]:
        """Analyze a single file for contextual markers and create entries."""
        entries = []
        relative_path = str(file_path.relative_to(base_path))
        
        try:
            with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                lines = f.readlines()
            
            for line_num, line in enumerate(lines, 1):
                line = line.strip()
                
                # Check if line has contextual significance
                if self._has_contextual_significance(line):
                    # Extract context from surrounding lines
                    context = self._extract_context(lines, line_num - 1, 2)
                    
                    # Create entry
                    entry = TagMapEntry(
                        document=relative_path,
                        line=line_num,
                        context=context,
                        preview=self._extract_preview(line),
                        title=self._extract_title_from_line(line),
                        date=self._extract_date_from_line(line, file_path),
                        category=self._determine_category(line),
                        tags=self._extract_tags(line)
                    )
                    
                    entries.append(entry)
                
                # Also check for marker patterns
                elif self._should_mark_line(line):
                    context = self._extract_context(lines, line_num - 1, 1)
                    
                    entry = TagMapEntry(
                        document=relative_path,
                        line=line_num,
                        context=context,
                        preview=self._extract_preview(line),
                        title=self._extract_title_from_line(line),
                        date=self._extract_date_from_line(line, file_path),
                        category=self._determine_category(line),
                        tags=self._extract_tags(line)
                    )
                    
                    entries.append(entry)
        
        except Exception as e:
            logger.error(f"Error analyzing file {file_path}: {e}")
        
        return entries
    
    def _extract_context(
        self, 
        lines: List[str], 
        line_number: int, 
        context_lines: int
    ) -> str:
        """Extract context from surrounding lines."""
        start = max(0, line_number - context_lines)
        end = min(len(lines), line_number + context_lines + 1)
        
        context_lines_list = lines[start:end]
        return ''.join(context_lines_list).strip()
    
    def _has_contextual_significance(self, line: str) -> bool:
        """Check if a line has contextual significance."""
        if not line.strip():
            return False
        
        # Check against contextual patterns
        for pattern in self.CONTEXTUAL_PATTERNS:
            if pattern.search(line):
                return True
        
        return False
    
    def _should_mark_line(self, line: str) -> bool:
        """Check if a line should be marked based on patterns."""
        if not line.strip():
            return False
        
        # Check against marker patterns
        for pattern in self.MARKER_PATTERNS:
            if pattern.search(line):
                return True
        
        return False
    
    def _build_cross_references(self, entries: List[TagMapEntry]):
        """Build cross-references between related entries."""
        # Create lookup for entries by document and line
        entry_lookup = {}
        for entry in entries:
            key = f"{entry.document}:{entry.line}"
            entry_lookup[key] = entry
        
        # Find related entries based on content similarity
        for entry in entries:
            if not entry.context:
                continue
            
            related = []
            entry_words = set(entry.context.lower().split())
            
            for other_entry in entries:
                if other_entry == entry:
                    continue
                
                if not other_entry.context:
                    continue
                
                other_words = set(other_entry.context.lower().split())
                
                # Calculate similarity
                intersection = entry_words.intersection(other_words)
                union = entry_words.union(other_words)
                
                if union:
                    similarity = len(intersection) / len(union)
                    if similarity > 0.3:  # 30% similarity threshold
                        related.append(f"{other_entry.document}:{other_entry.line}")
            
            entry.related_entries = related[:5]  # Limit to 5 related entries
    
    def _extract_preview(self, line: str, max_length: int = 100) -> Optional[str]:
        """Extract a preview from a line."""
        if not line.strip():
            return None
        
        preview = line.strip()
        if len(preview) > max_length:
            preview = preview[:max_length] + "..."
        
        return preview
    
    def _extract_title_from_line(self, line: str) -> Optional[str]:
        """Extract title from a line."""
        match = self.TITLE_PATTERN.search(line)
        if match:
            return match.group(1).strip()
        
        # Check for AmandaMap entry pattern
        match = self.AMANDAMAP_PATTERN.search(line)
        if match:
            return match.group(2).strip()
        
        return None
    
    def _extract_date_from_line(self, line: str, file_path: Path) -> Optional[str]:
        """Extract date from a line or file path."""
        # Check line for date pattern
        match = self.DATE_PATTERN.search(line)
        if match:
            return match.group(1)
        
        # Check file path for date
        filename = file_path.name
        match = self.DATE_PATTERN.search(filename)
        if match:
            return match.group(1)
        
        return None
    
    def _determine_category(self, line: str) -> Optional[str]:
        """Determine the category of content based on keywords."""
        line_lower = line.lower()
        
        for category, keywords in self.CATEGORY_KEYWORDS.items():
            if any(keyword in line_lower for keyword in keywords):
                return category
        
        return None
    
    def _extract_tags(self, line: str) -> List[str]:
        """Extract tags from a line."""
        tags = []
        line_lower = line.lower()
        
        # Extract hashtags
        hashtag_pattern = re.compile(r'#(\w+)')
        for match in hashtag_pattern.finditer(line):
            tags.append(match.group(1))
        
        # Extract keywords as tags
        for category, keywords in self.CATEGORY_KEYWORDS.items():
            for keyword in keywords:
                if keyword in line_lower:
                    tags.append(keyword)
        
        # Remove duplicates and return
        return list(set(tags))
    
    def load_tagmap(self, folder_path: Path) -> List[TagMapEntry]:
        """Load existing tagmap from file."""
        tagmap_path = folder_path / "tagmap.json"
        
        if not tagmap_path.exists():
            return []
        
        try:
            with open(tagmap_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            entries = []
            for entry_data in data:
                entry = TagMapEntry(**entry_data)
                entries.append(entry)
            
            return entries
            
        except Exception as e:
            logger.error(f"Error loading tagmap: {e}")
            return []
    
    def update_tagmap(
        self, 
        folder_path: Path, 
        progress_callback=None
    ) -> List[TagMapEntry]:
        """Update existing tagmap with new files."""
        logger.info(f"Updating tagmap for folder: {folder_path}")
        
        # Load existing entries
        existing_entries = self.load_tagmap(folder_path)
        existing_files = {entry.document for entry in existing_entries}
        
        # Generate new entries
        new_entries = self.generate_tagmap(folder_path, overwrite_existing=False, progress_callback=progress_callback)
        
        # Filter out entries for files that already exist
        new_entries = [entry for entry in new_entries if entry.document not in existing_files]
        
        # Combine and save
        all_entries = existing_entries + new_entries
        
        tagmap_path = folder_path / "tagmap.json"
        try:
            with open(tagmap_path, 'w', encoding='utf-8') as f:
                json.dump([asdict(entry) for entry in all_entries], f, indent=2)
            logger.info(f"Updated tagmap saved to {tagmap_path}")
        except Exception as e:
            logger.error(f"Error saving updated tagmap: {e}")
        
        return all_entries
    
    def get_tagmap_stats(self, entries: List[TagMapEntry]) -> Dict[str, Any]:
        """Get statistics about the tagmap."""
        categories = {}
        tags = {}
        
        for entry in entries:
            if entry.category:
                categories[entry.category] = categories.get(entry.category, 0) + 1
            
            for tag in entry.tags:
                tags[tag] = tags.get(tag, 0) + 1
        
        return {
            "total_entries": len(entries),
            "categories": categories,
            "top_tags": sorted(tags.items(), key=lambda x: x[1], reverse=True)[:10],
            "documents": len(set(entry.document for entry in entries if entry.document))
        } 